lib/actions/deploy.coffee

```diff
@@ -0,0 +1,168 @@
+
+	resin.models.application.get(appName)
+	.then (app) ->
+		console.log("Deploying #{hash} to fleet...")
+		resin.pine.patch
+			resource: 'application'
+			id: app.id
+			body:
+				commit: hash
+
+getBundleInfo = (options) ->
+	Promise = require('bluebird')
+	helpers = require('../utils/helpers')
+
+	Promise.try(-> helpers.getAppInfo(options.appName))
+	.then (app) ->
+		[app.arch, app.device_type]
+
+module.exports =
+	signature: 'deploy <appName> [image]'
+	description: 'Deploy a container to a resin.io application'
+	help: '''
+		Use this command to deploy and optionally build an image to an application.
+
+		Usage: deploy <appName> ([image] | -b [-c build-dir])
+
+		Note: If building with this command, all options supported by `resin build`
+		are also support with this command.
+
+		Examples:
+		$ resin deploy myApp -b -c myBuildDir/
+		$ resin deploy myApp myApp/myImage
+	'''
+	permission: 'user'
+	options: [
+		{
+			signature: 'build'
+			boolean: true
+			description: 'Build image then deploy'
+			alias: 'b'
+		},
+		{
+			signature: 'context'
+			parameter: 'context'
+			description: 'The context to use when building the image'
+			alias: 'c'
+		}
+	]
+	action: (params, options, done) ->
+		Promise = require('bluebird')
+		request = require('request')
+		tmp = Promise.promisifyAll(require('tmp'))
+		resin = require('resin-sdk-preconfigured')
+
+		dockerUtils = require('../utils/docker')
+
+		# Ensure the tmp files gets deleted
+		tmp.setGracefulCleanup()
+
+		docker = dockerUtils.getDocker(options)
+		new Promise (resolve, reject) ->
+			# Check input parameters
+			Promise.try( -> parseInput(params, options))
+			.then ([appName, build, context, imageName]) ->
+				tmp.tmpNameAsync()
+				.then (tmpPath) ->
+
+					# Setup the build args for how the build routine expects them
+					options.appName = appName
+					params.context = context
+					if build
+						# Instead of an image, the user has specified to build an image
+						imagePromise = dockerUtils.runBuild(params, options, getBundleInfo)
+					else
+						imagePromise = Promise.resolve(imageName)
+
+					imagePromise
+					.then (imageName) ->
+						Promise.all([
+							dockerUtils.bufferImage(docker, imageName, tmpPath)
+							resin.auth.getToken()
+							resin.auth.whoami()
+							resin.settings.get('resinUrl')
+							dockerUtils.getImageSize(docker, imageName)
+						])
+						.then ([stream, token, username, url, size]) ->
+							url = url || process.env.RESINRC_RESIN_URL
+							post = request.post
+								url: getBuilderPushEndpoint(url, username, params.appName)
+								auth:
+									bearer: token
+								body: stream
+
+							post
+							.on('error', reject)
+							.on 'data', (data) ->
+								handleMessage(data.toString())
+
+							# Set up upload reporting
+							pushProgress(size, post)
+
+							handleMessage = (data) ->
+								if process.env.DEBUG
+									console.log("Received data: #{data}")
+								obj = JSON.parse(data)
```

Comment:
`request` can if I set `json: true` in the request options, but then `request` tries to parse the image on the way to the builder, which is actually a tar archive so it fails. If you know how I can get `request` to only parse on the way back that would be much better.